keps/sig-network/3458-dynamically-re-configure-lb-node-set/README.md

```diff
@@ -0,0 +1,787 @@
+9. `Node2` is deleted
+10. KCCM re-configures LB set with all nodes
+11. LB is now correctly configured.
+
+Given this: the KCCM will experience transient errors between 8. and 10. Keeping
+that condition might thus make sense, from the point of view of avoiding that
+transient failure mode in the KCCM.
+
+[The auto-scaler performs down-scaling in two
+modes](https://github.com/kubernetes/autoscaler/blob/8dcd3b3eb2d1d28b7ba05a3af48a171cb40547b9/cluster-autoscaler/core/scaledown/actuation/actuator.go#L72-L106):
+
+- a) either by deleting empty nodes which are not running any pods, in which case
+  it taints the Node and asynchronously deletes the instance.
+- b) or by deleting a non-empty node, in which case it taints the Node and only
+  after that: drains it and deletes the instance.
+
+Removing the Node from the load balancer set does not matter for a) since no
+pods are running on the Node: there are no connections to drain. For b),
+removing the Node from the set will cause all connections to get terminated
+instantly.
+
+Given this: it would seem preferable to remove the taint condition from the KCCM
+and only delete the node when receiving the delete event for the Node. However,
+it might be better to give more weight to the avoiding errors in the KCCM than
+optimally connection draining nodes which are simultaneously being drained of
+endpoints. Hence, keeping the condition in the KCCM, as-is, might be the safest
+path forward.
+
+The second predicate has no direct correlation to networking, but __could__,
+__possibly__, __maybe__ indicate a general problem with the node should the
+Kubelet be un-healthy in certain specific cases (for ex: if it's resource
+starved, which could be impacting other system processes as well, and
+subsequently networking related processes).
+
+We can define two classes of services, when the type is `LoadBalancer`.
+
+- A) services which have `externalTrafficPolicy: Cluster`
+- B) services which have `externalTrafficPolicy: Local`
+
+These are the only two classes of services which impact traffic forwarding for a
+load balancer and answer the question: "where should traffic be sent to?".
+
+Load balancers, today, go about figuring out where "traffic should be sent to"
+through the utilization of an HC. This is the primary API offered by clouds,
+which allows indicating if nodes should be used by a load balancer, or not. The
+HC is configured differently, by Kubernetes, depending on the class of service.
+
+- For services of class B): the HC is configured to probe all nodes in its set
+  on a node port defined to be value of the field `healthCheckNodePort`. The
+  field can be user defined, but when it's not: it's automatically generated by
+  Kubernetes. The entity answering to the HC on this port is the service proxy
+  and its answer currently reflects one sole criteria: does the service have a
+  __ready__ endpoint running on the node in question? If the answer is "yes",
```

Comment:
This KEP only focuses on health from the perspective of the Node whereas ProxyTerminatingEndpoints focuses on the Pod. There is a subtle link though, because the readiness of the Node might impact the readiness of the Pod if the Node stays `NotReady` for a sufficiently long time whereby the kubelet (or API-server, I misremember) will mark all Pods on the Node `NotReady` too.

Reply:
Your last point is what I was trying to say.